### ISSUE TITLE
Storage order? (col major vs row major)

### ISSUE THREAD

**Issue by NeumannIMT** [MagneticParticleImaging/MDF#74]
has i was working on #70 i noticed that storage order was somehow an issue. 

The problem: 
MATLAB           is col major
Julia                  is col major
C                      can be consider being row major
HDF5                is row major
(Eigen3 C++)    is col major (default; can be switched to row major [row major is not tested that well-> see doc])
Python              is row major (default; can be switched between c (row) and fortran (col) ordering) 
[I misinterpreted the documentation of python in the 30 min i used learning a bit of python] 

So my question is: 
Do we care about the fact that data written to hdf5 is transposed in the viewer if written directly from Matlab or Julia? If we read those files into python it might be a problem. 

*edit: correct

**Comment by tknopp** [comment]
* Python (numpy) is certainly not col major but row major.
* Eigen can be both
* C is actually nothing because it has no support for multidimensional (dynamic! heap!) arrays.

**Comment by tknopp** [comment]
But to answer the question: We use the hdf5 convention in the specification. The dimensions are by purpose arranged in a way that cache is utilized best. sequential data is also stored sequentially.

When using a certain programming language one has to take all this into account. But there is nothing we can actually do about that.

**Comment by profix898** [comment]
We are using HDF.PInvoke wrapper for our C# implementation and there was a discussion regarding row- vs. column-major in HDF5 a while ago, see https://github.com/HDFGroup/HDF.PInvoke/issues/52
Basically I always considered Julia as the reference implementation simply because HH is doing it and is also master of MDF. So we adjusted our C++/C# implementation to flip dimensions accordingly. We should probably add a note the MDF spec to clarify, but otherwise I also dont see what we could do more.

**Comment by tknopp** [comment]
I am not sure I understand where you had to flip the dimensions. The MDF should not contain anything that is specific to Julia or HH working group.

Are you talking about the way how the system matrix is stored?

**Comment by profix898** [comment]
Did you read [7.3.2.5. C versus Fortran Dataspaces](https://support.hdfgroup.org/HDF5/doc/UG/HDF5_Users_Guide-Responsive%20HTML5/index.html#t=HDF5_Users_Guide%2FDataspaces%2FHDF5_Dataspaces_and_Partial_I_O.htm%23TOC_7_3_2_Dataspacebc-5&rhtocid=7.1.0_2)? Storage order (esp. system matrices) in MDF make sense. Still dimensions written from Julia/Matlab (Fortran-style languages) appear flipped if you read them from C-style languages. You can easily verify by opening a dataset in HDFView. So this in only a documentation issue.

**Comment by tknopp** [comment]
yes sure, but this has nothing to do with MDF. If I read any HDF5 file in Matlab, the dimensions are always flipped.

**Comment by tknopp** [comment]
(we might should explain this in the documentation, since this might not be clear for people new to HDF5)

**Comment by hofmannmartin** [comment]
will do so tomorrow

**Comment by NeumannIMT** [comment]
> Python (numpy) is certainly not col major but row major. Eigen can be both  C is actually nothing because it has no support for multidimensional (dynamic! heap!) arrays.

I edited my post to be correct. But dont get stuck on details with that please.
let me reformulate my question a bit:

**How would you store a system matrix in an normal matrix?**
I would store all the frequencies components (K) of one position in a rowwise fashion and store subsequent new positions (N) in another row. This means if i select one column I get one frequency component and all spatial positions. Leading to an K x N matrix (I believe that is the expected ordering if people think about system matrices)

Depending of the storage order of my matrix either the column or the row will be the fast access dimension and i believe that accessing the column should be the fast direction. 

**So how does that translate to MDF?**
MDF is designed to have the dimension N x K (Ignoring J and C) this means K ist the fast direction but it will be saved in the row direction (Meaning that the column major matrix was transposed in HDF5 by directly writing it to the file without consideration that it should by in the expected form).  
But then there is the isPermutedFlag in the specification which allows to permuted N and K to be K x N which results in a matrix which is in the expected form for a system matrix and N being the fast direction. 

**Is this the intend behind the isPermutedFlag? I always wondered why it is even there?**

**Now to the problem with generic programming:** 
If I have a n x m matrix in my programm and want to write it to HDF5 the user expects the data in the file to be also n x m independend of storage direction. What MDF currently does by **default** is transpose the matrix n x m into m x n due to the column major storage layout and thats why i asked this question  (should this be default behavior?)

**Comment by hofmannmartin** [comment]
The MDF should be independent of a specific implementation. The last dimension in the specification should be the fastest to access and the first the slowest. So specifications refer to the storage order inside the HDF5 file.

 Depending on how your programming language lays out the memory that means that if you load the data the dimensions will be reversed (col major) with respect to the specifications. But in any case the fast dimension stays the fast one and the slow stays the slow one, which is the important point (see below why).

> Is this the intend behind the isPermutedFlag? I always wondered why it is even there?

The intend behind the flag is the possibility to get system matrices ready for reconstruction as fast as possible. Often a frequency selection is performed, in which case it is advantageous to have the dimension CxK being slow to access, since in that case not the whole data has to be transversed to perform the selection. Furthermore there are reconstruction algorithms which require fast access to the N dimension. In this case it is beneficial to have the data already ready in the right storage order, instead of having to reorder 20GB or so each time you start a reconstruction. Especially in a online reconstruction scenario, where you switch system matrices from time to time. Of cause for measurements you usually want to have the the data of a single frame contiguous in the memory and not distributed throughout the whole file.

So the only satisfactory solution is to allow both storage orders either frequency or position beeing fast to access.

**Comment by hofmannmartin** [comment]
The issue is now documented in the specifications, which will cause that the ordering of the dimensions of multidimensional arrays will depend on the programming language, which is used.

**Comment by NeumannIMT** [comment]
@hofmannmartin: Considering your answer is the flag "isPermutedFlag" a good name for what it is intended for? Meaning conserving fast storage access? 
For this we would need information about the order of the incoming order since it could be arbitarly ordered. (Also I agree that not all ordering dimension would make sense)

So the question boils down what would a user expect from a system matrix stored in a hdf5 file?
I think the user would expect a matrix which is K x N. So if it is N x K in the file would it not make more 
sense to change the flag to isTransposed and only use two dimensions for a system matrix? (From a user perspective that is what is expected from a system matrix or not?)
 (I intepret K aus Sum(K_C) and no patches J since these are in my opinion seperate files -> maybe v3?)
(can somebody link me to the discussion for N x J x C x K ? This seems like a really bad ordering choice. Shouldn't it be J x N x C x K )

**Comment by hofmannmartin** [comment]
> For this we would need information about the order of the incoming order since it could be arbitarly ordered. (Also I agree that not all ordering dimension would make sense)

Sorry, but I do not get what you are trying to say.

> So the question boils down what would a user expect from a system matrix stored in a hdf5 file?
I think the user would expect a matrix which is K x N. So if it is N x K in the file would it not make more
sense to change the flag to isTransposed and only use two dimensions for a system matrix? (From a user perspective that is what is expected from a system matrix or not?)

@NeumannIMT: If we would only have two dimensions I would agree, but we have more and what we do is not just a simple transposition. `isPositionsFastToAccess` is an alternative proposal. What do you think?

> (I intepret K aus Sum(K_C) and no patches J since these are in my opinion seperate files -> maybe v3?)
(can somebody link me to the discussion for N x J x C x K ? This seems like a really bad ordering choice. Shouldn't it be J x N x C x K )

The ordering choice is actually not so bad. Consider you want to reconstruct a single measurement n. Then you would have to go through the whole file to gather the information for all patches if J is slowest to access. The current order allows you to easily grab the data obtained for all patches of a measurement since it is stored contiguously.

As for the proposal to split the data into separate files. That might be an option for small J, but consider a Cartesian measurement where you scan maybe 1000 lines (patches) or more. The data from each line would have to be stored in a single file. That would in the end also result slow performance if you want to read a single measurement, since the data is distributed throughout all of the files.

At HH we are handling multi patch data quite frequently and kept the the data acquired at different patch positions (making a fame) together.

**Comment by tknopp** [comment]
>  (I intepret K aus Sum(K_C) and no patches J since these are in my opinion seperate files -> maybe v3?)
(can somebody link me to the discussion for N x J x C x K ? This seems like a really bad ordering choice. Shouldn't it be J x N x C x K )

You seem to confuse things here N x J x C x W is exactly the ordering how the data is measured. With one frame (enumerated by N) there are J x W measurement points. And W is faster than J. Since measurements are taken in parallel the C dimension is the second fastest.

> So the question boils down what would a user expect from a system matrix stored in a hdf5 file?
I think the user would expect a matrix which is K x N. So if it is N x K in the file would it not make more
sense to change the flag to isTransposed and only use two dimensions for a system matrix? (From a user perspective that is what is expected from a system matrix or not?)

The field `/measurement/data` is not the system matrix (in general). It is the measurement data and we store the data exactly in the order how it has been measured with the receiver. It is possible to apply various transformations to bring the data in a suitable format for deriving the system matrix from it. This is where all the flags are for.

The memory representation and order is (by the way) exactly the same as that stored within the Bruker data format. Also the IMT HDF internal dataformat uses the same ordering (though patches are not included IMHO). The order has also been reviewed by @profix898, by colleagues from the Würzburg group, by J. Rahmer (who has by far the most experience with Multi-Patching). This does not imply that be don't do errors but I want to express that we have put serious effort to ensure a sensible ordering of the data.

**Comment by NeumannIMT** [comment]
As I recieved your response i was thinking: 
"Why did I propose the J x N x C x K ordering?"  at that time it felt right but I couldn't find the argument because the N x J x C x K also suddenly felt correct. (and in some post i even wroten that N x J x ANY is obviously correct).
So i was thinking what was it that drove me to propose J x N x C x K? And the answer to that is: patch and frame ordering: You can either acquire N frames of a patch and then move to another patch (which is J x N x C x K) or you can acquire J patches and then move to the next frame (N x J x C x K). Due to timing/acqusition reasons you are more likly to want  N x J x C x K than J x N x C x K but both are valid acquisition schemes
 
This basically brings me back to the questions raised in #72 about frames and patches. 
MDF needs a FAQ for such things. 

>@NeumannIMT: If we would only have two dimensions I would agree, but we have more and what we do is not just a simple transposition. isPositionsFastToAccess is an alternative proposal. What do you think?

I am currently thinking about how to support the behavior from a writing position. Because in general if I have a certain matrix layout I would also like to have the same matrix layout in the HDF5 file. I would need an option like: `ignoreStorageLayout` in my generic implementation of HDF5 to directly write it into the HDF5 file without transposing the data.  How to handle it in MDF? Hmm maybe we want to store information about the writer in the file, like if the code was col or row major (feels like a bad idea)?
Ok lets move one step back. What we want: Fast access in all cases. 
**Case acquisition:** Data should be saved as it comes in so (N x J x) C x W makes sense 
(although channel data may be interleaved; I am strongly for ordering it)
Is there a good case to allow  J x C x W x N here? I dont see it. Data will always come has C x W (or C x K) packages so why allow the permutation here?. 
**Case calibration:** Data should be saved as it comes in which will be in C x W (or C x K)  packages
**Case measurement data is a system matrix** The only case where it makes sense to have the permutation in my opinion. My conclusion would be that measurement/data/ is not the right place to store the systemmatrix. But I understand that the storage here is a convenient place. Maybe the option should thus be called `isSystemMatrixLayout` which makes clear that the data has been reorder to allow fast readin of one frequency component of the system matrix? 
If I am missing some other use case for the permutation please tell me. 

Considering access speed a small note: We did not yet considered chuncked and/or compress datasets.

**Comment by tknopp** [comment]
`J x C x W x N` is a system matrix in time domain. Since `x-space` is also in our focus, we allow this.

**Comment by NeumannIMT** [comment]
@tknopp thats an argument for `isSystemMatrixLayout`. The permutation seems to only makes sense if measurement/data should be interpreted as a system matrix

**Comment by tknopp** [comment]
`Case calibration` and `Case measured data is a system matrix` are similar. `Case calibration` is the "raw" form. And one can use this for reconstruction (if reordering is done on the fly). `Case measurement data is a system matrix` is when several post processing steps are applied. Calling this `isSystemMatrixLayout` would be an option. But keep in mind that this is actually not yet the system matrix. It also contains background measurements.

**Comment by tknopp** [comment]
(And are we really talking just about the name of this parameter?)

**Comment by tknopp** [comment]
I am against `isPermuted`. `isTransposed` is not bad IMHO. `isFastFrameAxis` could be an option.

`isSystemMatrixLayout` wrongly assumes that it is already the system matrix. Also in time domain, the transposition makes not so much sense. So there `N x J x C x W` would be a valid choice for a system matrix. `isFrameAxisPermuted` is another option.

**Comment by hofmannmartin** [comment]
> As I recieved your response i was thinking:
"Why did I propose the J x N x C x K ordering?" at that time it felt right but I couldn't find the argument because the N x J x C x K also suddenly felt correct. (and in some post i even wroten that N x J x ANY is obviously correct).
So i was thinking what was it that drove me to propose J x N x C x K? And the answer to that is: patch and frame ordering: You can either acquire N frames of a patch and then move to another patch (which is J x N x C x K) or you can acquire J patches and then move to the next frame (N x J x C x K). Due to timing/acqusition reasons you are more likly to want N x J x C x K than J x N x C x K but both are valid acquisition schemes

I think this can be resolved by semantics. A measurement where you acquire M measurements at a certain patch position and then moving to the next one of the P positions and so on, is simply a measurement, where you have J=MP. There is no convention that forbids you to have the same patch position M times in a row. Making this choice simply eliminates the case that you have brought up, where the order of J and N is not NxJ.

A frame is the smallest collection of measurements, which is periodic. This excludes cases, where you would measure one frame here and the next there. But that for sure needs to be stated within the MDF.

> This basically brings me back to the questions raised in #72 about frames and patches.
MDF needs a FAQ for such things.

This is in principle a good idea. We would need someone asking the questions.

**Comment by tknopp** [comment]
`numPeriods`: `Number of drive-field periods per patch`

`numSamplingPoints`:   `points sampled at receiver during one patch (product of drive field period, number of periods, and number of averages)`

**Comment by tknopp** [comment]
So a frame may contain several periods. which addresses @NeumannIMT use case. Actually this will be a common case.

**Comment by NeumannIMT** [comment]
> (And are we really talking just about the name of this parameter?)

Yes, because sometimes it is important to have good names which convey meaning. isPermuted can mean everything it does not tell you what has been permutated (meaning which dimensions) an why. Whereas `isSystemMatrixLayout` or even `isSystemMatrix` conveys meaning that the data is stored _to be used_ as a SystemMatrix. 

> `isSystemMatrixLayout` wrongly assumes that it is already the system matrix.

why do you assume that? It should just say to be used as a (part) system matrix but maybe we need a more verbose name like `isFastkSpaceSystemMatrixLayout` (because lets be realistic thats currently the only use case)

`isTransposed` does not convey meaning beyond a 2D matrix as @hofmannmartin pointed out.

`isFastFrameAxis` My question would be what is the FrameAxis? Its due to the double meaning of N which can be either a Frame number or a position. Maybe: `hasFastPositionAccess` 

> Also in time domain, the transposition makes not so much sense.

Thought so too but I am not a reconstruction guy ;)

> We would need someone asking the questions.

Am I not already doing that? I feel like im nagging and complaining all the time.

> So a frame may contain several periods. which addresses @NeumannIMT use case. Actually this will be a common case.

do i understand correctly that J x N' x C x K would in reallity be something like 1 x J x C x K with numPeriods = N'*numPeriods' (I do not know if that is the wanted behavior but i also dont do anything with patches so i cannot tell)
 
> A patch contains a single drive field period

No it will be numPeriods as far as i understood it

Additional to the FAQ we need a List of covered cases

**Comment by tknopp** [comment]
> isFastFrameAxis My question would be what is the FrameAxis? Its due to the double meaning of N which can be either a Frame number or a position. Maybe: hasFastPositionAccess

`N` has no double meaning. It is the number of frames (see `numFrames`). The positions may correspond to frames, but this is not always the case. The IMT for instance acquires data that has no positions, but OffsetFields.

**Comment by NeumannIMT** [comment]
From the MDF spec:
> N acquired frames, same as a spatial position for calibration

Maybe that is a bug in the spec. I just realized that your are storing background frames and the delta sample data in the same dataset N-O = BG Frames. 
Is there a good reason to not have a dedicated ready to use dataset for the system matrix? 
(except for storage considerations)

**Comment by hofmannmartin** [comment]
> do i understand correctly that J x N' x C x K would in reallity be something like 1 x J x C x K with numPeriods = N'*numPeriods' (I do not know if that is the wanted behavior but i also dont do anything with patches so i cannot tell)

No. J' x N' x C x K would be something like N x J x C x K, where N=1 and J=J'*N'

> No it will be numPeriods as far as i understood it

correct.

> `N` has no double meaning. It is the number of frames (see `numFrames`). The positions may correspond to frames, but this is not always the case. The IMT for instance acquires data that has no positions, but OffsetFields.

In that sense it always has the meaning of a frame, but can have additional meaning depending on the context.

Therefore `isFrameAxisFast` would be an appropriate name for the field.

**Comment by tknopp** [comment]
> Is there a good reason to not have a dedicated ready to use dataset for the system matrix?
(except for storage considerations)

yes. we need that. Ready to use system matrices can be stored (so its possible), but it is also possible to store the system matrix before BG correction. Explaining why is difficult but there are situations where the measurement data has no background measurement and then it is handy to use that measured in the system matrix. Further the image quality is sometimes superior when applying no BG correction to the system matrix.

**Comment by NeumannIMT** [comment]
> No. J' x N' x C x K would be something like N x J x C x K, where N=1 and J=J'*N'

got the idea

> Therefore isFrameAxisFast would be an appropriate name for the field.

Ok. Maybe `hasFastFrameAccess` is better. `isSomething` does not feel right. The reason for that is that the other `is` flags are flags which described how the data has been changed (e.g. background correction, transfer function etc.) whereas this just is a reordering of the data itself. Under this viewpoint I would also change
`isFramePermution`  (the typo has been copied from the spec.) to `hasFramePermutation`

The suggestion would be to use `isSomething` for properties describing the data itself and `hasSomething` for properties describing the dataset and not the data. (`has` flags are not allowed to change the data only reorder it)

> Ready to use system matrices can be stored (so its possible)

from my viewpoint this feels like a hack if the data has been heavily postprocessed (from my viewpoint it doesn't really fit into /measurement/data anymore). I will open another issue to discuss this to get other opinions (seems to be an extension to mdf and not a braking change)

Lets just end the other discussions here and decide how to name the flags.

**Comment by hofmannmartin** [comment]
> The suggestion would be to use `isSomething` for properties describing the data itself and `hasSomething` for properties describing the dataset and not the data. (has flags are not allowed to change the data only reorder it)

Our idea was to use the `is` as an indication that the field contains some Boolean data. I think we have this consistent throughout the whole document and I would rather keep it that way.

**Comment by tknopp** [comment]
> from my viewpoint this feels like a hack if the data has been heavily postprocessed (from my viewpoint it doesn't really fit into /measurement/data anymore). I will open another issue to discuss this to get other opinions (seems to be an extension to mdf and not a braking change)

Look at previous versions of the specification (around May/June). We had a processing group in the mean time but removed it because it was heavily redundant with the measurement group. Our system now is fully transparent. We store the data and tell the user, which processing have been applied.

**Comment by NeumannIMT** [comment]
> Our idea was to use the `is` as an indication that the field contains some Boolean data. I think we have this consistent throughout the whole document and I would rather keep it that way.

Dont see the issue, it is a new field. Both `isSomething` and `hasSomething` indicate that the field contain boolean data. But I have to say I also just like the syntatic sugar `has` would add in conveying meaning. 
So you would have: 
`isSomething` : Be carefull something has been done to the data directly which might screw up your reco
`hasSomething` : Be carefull something has been done with the order of the data which might screw up your reco

maybe it just comes down to personal taste and style? seems like a good point to get more opinions and vote for the one or other. 

> Look at previous versions of the specification (around May/June). We had a processing group in the mean time but removed it because it was heavily redundant with the measurement group. Our system now is fully transparent. We store the data and tell the user, which processing have been applied.

seen that. havent found any good solution yet (was thinking of changing measurement/data to data/data or adding a flag to calibration that the data is saved in measurement/data but found that stupid because it needs a string/definition somewhere where the actual data is saved). 

What are the options:
dropping data into / -> Will litter / or we really need attributes for the dataset to not litter / too much
allow <contextdependentgroupname>/data? -> will make programming more verbose although it could reduce error reading the wrong file. (Wait you wanted to load a system matrix not the measurement)
just rename it to moregeneralname/data? -> difficult to find more general name, also loses its natural group for simple measurements. (Maybe storage/data ?)

I am somewhat ok with measurement/data but also not 100% happy about it.

**Comment by hofmannmartin** [comment]
> Dont see the issue, it is a new field. Both isSomething and hasSomething indicate that the field contain boolean data. But I have to say I also just like the syntatic sugar has would add in conveying meaning.
So you would have:
isSomething : Be carefull something has been done to the data directly which might screw up your reco
hasSomething : Be carefull something has been done with the order of the data which might screw up your reco
>
> maybe it just comes down to personal taste and style? seems like a good point to get more opinions and vote for the one or other.

Would you please open another issue for that, since it is not really the point of this one.

**Comment by tknopp** [comment]
I changed to `isFastFrameAxis`. Naming bikeshed is still open but can happen in #88.
